rust/cubesqlplanner/cubesqlplanner/src/planner/time_dimension/date_time.rs

```diff
@@ -0,0 +1,390 @@
+        let adjusted_date = NaiveDate::from_ymd_opt(year, month as u32, day)
+            .or_else(|| {
+                (1..=31)
+                    .rev()
+                    .find_map(|d| NaiveDate::from_ymd_opt(year, month as u32, d))
+            })
+            .ok_or_else(|| {
+                CubeError::internal(format!(
+                    "Failed to compute valid date while adding interval {:?} to date {}",
+                    interval, self.date_time
+                ))
+            })?;
+
+        // Step 2: Add weeks and days
+        let adjusted_date =
+            adjusted_date + Duration::days(interval.week as i64 * 7 + interval.day as i64);
+
+        // Step 3: Recombine with original time
+        let time = self.naive_local().time();
+        let mut naive = NaiveDateTime::new(adjusted_date, time);
+
+        // Step 4: Add time-based parts
+        naive = naive
+            + Duration::hours(interval.hour as i64)
+            + Duration::minutes(interval.minute as i64)
+            + Duration::seconds(interval.second as i64);
+
+        Self::from_local_date_time(self.date_time.timezone(), naive)
+    }
+
+    pub fn sub_interval(&self, interval: &SqlInterval) -> Result<Self, CubeError> {
+        self.add_interval(&interval.inverse())
+    }
+
+    pub fn granularity(&self) -> String {
+        let time = self.date_time.time();
+
+        let weekday = self.date_time.weekday();
+        let is_zero_time = |t: chrono::NaiveTime| {
+            t.hour() == 0 && t.minute() == 0 && t.second() == 0 && t.nanosecond() == 0
+        };
+
+        if self.date_time.month() == 1 && self.date_time.day() == 1 && is_zero_time(time) {
+            "year".to_string()
+        } else if self.date_time.day() == 1 && is_zero_time(time) {
+            "month".to_string()
+        } else if weekday == Weekday::Mon && is_zero_time(time) {
+            "week".to_string()
+        } else if is_zero_time(time) {
+            "day".to_string()
+        } else if time.minute() == 0 && time.second() == 0 && time.nanosecond() == 0 {
+            "hour".to_string()
+        } else if time.second() == 0 && time.nanosecond() == 0 {
+            "minute".to_string()
+        } else if time.nanosecond() == 0 {
+            "second".to_string()
+        } else {
+            "second".to_string()
+        }
+    }
+
+    pub fn align_to_origin(
+        &self,
+        origin: &Self,
+        interval: &SqlInterval,
+    ) -> Result<Self, CubeError> {
+        let mut aligned = self.clone();
+        let mut offset = origin.clone();
+
+        if self < origin {
+            while &offset > self {
+                offset = offset.sub_interval(interval)?;
+            }
+            aligned = offset;
+        } else {
+            while &offset < self {
+                aligned = offset.clone();
+                offset = offset.add_interval(interval)?;
+            }
+            println!("offset: {:?}, aligned: {:?}", offset, aligned);
```

Comment:
![image](https://github.com/user-attachments/assets/712b5b68-6281-4272-8ee0-40659c0688ca)

Reply:
You don't see it :) Close your eyes! Both! :)

Reply:
You didn't see that :) fixed